src/abstractdataframe/selection.jl

```diff
@@ -0,0 +1,357 @@
+│ 1   │ 1     │ 4     │
+│ 2   │ 2     │ 5     │
+│ 3   │ 3     │ 6     │
+
+julia> select(df, :b, :a=><(1.5)=>:c)
+3×2 DataFrame
+│ Row │ b     │ c    │
+│     │ Int64 │ Bool │
+├─────┼───────┼──────┤
+│ 1   │ 4     │ 1    │
+│ 2   │ 5     │ 0    │
+│ 3   │ 6     │ 0    │
+```
+
+"""
+select(df::DataFrame, inds::AbstractVector{Int}; copycols::Bool=true) =
+    DataFrame(_columns(df)[inds], Index(_names(df)[inds]),
+              copycols=copycols)
+select(df::DataFrame, c::Union{AbstractVector{<:Integer}, AbstractVector{Symbol},
+                               Colon, All, Not, Between, Regex}; copycols::Bool=true) =
+    select(df, index(df)[c], copycols=copycols)
+select(df::DataFrame, c::ColumnIndex; copycols::Bool=true) =
+    select(df, [c], copycols=copycols)
+
+function select(df::DataFrame, cs...; copycols::Bool=true)
+    newdf = DataFrame()
+    # this line ensures that we fail early if passed source column names are missing
+    ncs = [normalize_selection(index(df), c) for c in cs]
+    # it should be OK to be type unstable here + in this way we aviod having to compile custom Dict
+    transformed_cols = Dict()
+    for nc in ncs
+        if nc isa Pair
+            lnc = last(nc)
+            newname = lnc isa Symbol ? lnc : last(lnc)
+            @assert newname isa Symbol
+            if haskey(transformed_cols, newname)
+                throw(ArgumentError("duplicate target transformed or renamed column names passed"))
+            end
+            transformed_cols[newname] = nc
+        end
+    end
+    for nc in ncs
+        if nc isa Union{Int, AbstractVector{Int}}
+            allunique(nc) || throw(ArgumentError("duplicate column names selected"))
+            for i in nc
+                newname = _names(df)[i]
+                if !hasproperty(newdf, newname)
+                    if haskey(transformed_cols, newname)
+                        nct = transformed_cols[newname]
+                        @assert !isnothing(nct)
+                        select_transform!(nct, df, newdf, transformed_cols, copycols)
+                    else
+                        newdf[!, newname] = copycols ? df[:, i] : df[!, i]
+                    end
+                end
+            end
+        else
+            @assert nc isa Pair{Int, <:Union{Symbol, Pair{<:Function, Symbol}}}
+            select_transform!(nc, df, newdf, transformed_cols, copycols)
+        end
+    end
+    return newdf
+end
+
+select(dfv::SubDataFrame, ind::ColumnIndex; copycols::Bool=true) =
+    select(dfv, [ind], copycols=copycols)
+select(dfv::SubDataFrame, inds::Union{AbstractVector{<:Integer}, AbstractVector{Symbol},
+                                      Colon, All, Not, Between, Regex}; copycols::Bool=true) =
+    copycols ? dfv[:, inds] : view(dfv, :, inds)
+
+function select(dfv::SubDataFrame, inds...; copycols::Bool=true)
+    if copycols
+        return select(copy(dfv), inds..., copycols=false)
+    else
+        # we do not support transformations here
```

Comment:
Throw an error if `inds` contains transformations? The one `view` is going to throw will likely be quite cryptic.

Reply:
I have also thought about it. Now it throws:
```
julia> select(dfv, :a=>:b, copycols=false)
ERROR: MethodError: no method matching getindex(::DataFrames.Index, ::Pair{Int64,Symbol})
```
but the problem is that it adding a more sophisticated check here will not be extensible (if we ever change `getindex` for `Index` we will have to update the code here, and it is likely that it will be forgotten).

I assume we can safely take that `Pair` will never be allowed in `getindex`. I will just add this simple test that should catch this and it should cover a majority of user-errors I think.